### PR TITLE
Broker #4: Brokered Silent Requests

### PR DESCRIPTION
**Pull request by tnorling**
Adds support for brokered silent requests. When an app calls `acquireTokenSilent` MSAL will perform the cache lookup, then if a refresh is required it will either attempt to use it's refresh token to obtain a new token (existing scenario) or if the app has brokering enabled, will send a message up to the broker to perform the refresh for it (new scenario)

### CODE REVIEW

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 1]
```diff
@@ -373,50 +373,25 @@ export abstract class ClientApplication {
      * @returns {Promise.<AuthenticationResult>} - a promise that is fulfilled when this function has completed, or rejected if an error was raised. Returns the {@link AuthResponse} object
      *
      */
-    async acquireTokenSilent(request: SilentRequest): Promise<AuthenticationResult> {
+    protected async refreshToken(request: SilentRequest): Promise<AuthenticationResult> {
```
Comment (pkanher617) [inline lib/msal-browser/src/app/ClientApplication.ts:376]: change the name of this to `acquireTokenByRefreshToken`

File: lib/msal-common/src/client/RefreshTokenClient.ts [hunk 2]
```diff
@@ -43,7 +47,32 @@ export class RefreshTokenClient extends BaseClient {
         return tokenResponse;
     }
 
-    private async executeTokenRequest(request: RefreshTokenRequest, authority: Authority)
+    public async refreshToken(request: SilentFlowRequest): Promise<AuthenticationResult> {
```
Comment (pkanher617) [inline lib/msal-common/src/client/RefreshTokenClient.ts:50]: change to `acquireTokenByRefreshToken`

File: lib/msal-browser/src/broker/BrokerClientApplication.ts [hunk 3]
```diff
@@ -112,6 +115,19 @@ export class BrokerClientApplication extends ClientApplication {
         }
     }
 
+    private async brokeredSilentRequest(validMessage: BrokerAuthRequest, clientPort: MessagePort): Promise<void> {
+        try {
+            const response: BrokerAuthenticationResult = (await this.refreshToken(validMessage.request as SilentRequest)) as BrokerAuthenticationResult;
+            const brokerAuthResponse: BrokerAuthResult = new BrokerAuthResult(InteractionType.SILENT, response);
```
Comment (pkanher617) [inline lib/msal-browser/src/broker/BrokerClientApplication.ts:121]: Need to find some better names here - `BrokerAuthenticationResult` and `BrokerAuthResult` are a bit too similar.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-browser/src/app/ClientApplication.ts:376]: +1 Can be confusing otherwise!

File: lib/msal-browser/src/broker/BrokerClientApplication.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/src/broker/BrokerClientApplication.ts:121]: `BrokerAuthResponse` should be fine IMO

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-browser/src/app/ClientApplication.ts]: Done

File: lib/msal-common/src/client/RefreshTokenClient.ts (hunk 2 above)
Comment (tnorling) [reply lib/msal-common/src/client/RefreshTokenClient.ts]: Done

File: lib/msal-browser/src/broker/BrokerClientApplication.ts (hunk 3 above)
Comment (tnorling) [reply lib/msal-browser/src/broker/BrokerClientApplication.ts]: Done

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 4]
```diff
@@ -58,11 +58,9 @@ export class PublicClientApplication extends ClientApplication implements IPubli
         } else if (this.config.system.brokerOptions.allowBrokering) {
             this.embeddedApp = new EmbeddedClientApplication(this.config, this.logger, this.browserStorage);
             this.logger.verbose("Acting as child");
-            try {
-                this.embeddedApp.initiateHandshake();
-            } catch (e) {
+            this.embeddedApp.initiateHandshake().catch((e) => {
```
Comment (DarylThayil) [inline lib/msal-browser/src/app/PublicClientApplication.ts:61]: does this need to be awaited?

Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts:61]: It does, but not here. PR #2300 exposes this as a public API which we will expect people to call and await in their app rather than doing async stuff in the constructor

**Review by jmckennon — APPROVED** [review]
Looks good, I'd like to see tests (if they aren't in a different PR already) before merging though

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 5]
```diff
@@ -136,6 +134,9 @@ export class PublicClientApplication extends ClientApplication implements IPubli
             const silentAuthClient = await this.createSilentFlowClient(serverTelemetryManager, silentRequest.authority);
             return silentAuthClient.acquireCachedToken(silentRequest);
         } catch (e) {
+            if (this.embeddedApp && this.embeddedApp.brokerConnectionEstablished) {
```
Comment (sameerag) [inline lib/msal-browser/src/app/PublicClientApplication.ts:137]: Why are we not checking it before the catch instead of letting silent fail?

Comment (tnorling) [reply lib/msal-browser/src/app/PublicClientApplication.ts:137]: Tokens will be cached on the embedded app, so we only need to check if brokering is enabled before making the network request.

File: lib/msal-browser/src/broker/BrokerAuthResponse.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    static processBrokerResponse(brokerAuthResultMessage: MessageEvent, browserStorage: BrowserStorage): AuthenticationResult {
+        const brokerAuthResult = BrokerAuthResponse.validate(brokerAuthResultMessage);
+        if (brokerAuthResult.error) {
+            throw brokerAuthResult.error;
+        }
+        const accessTokenEntity: AccessTokenEntity = new AccessTokenEntity();
+        const idTokenEntity: IdTokenEntity = new IdTokenEntity();
+        const accountEntity: AccountEntity = new AccountEntity();
+        const cacheRecord: CacheRecord = {
+            accessToken: CacheManager.toObject(brokerAuthResult.result.tokensToCache.accessToken, accessTokenEntity) as AccessTokenEntity,
```
Comment (sameerag) [inline lib/msal-browser/src/broker/BrokerAuthResponse.ts:42]: too wordy? Temp variables?

File: lib/msal-browser/src/broker/EmbeddedClientApplication.ts [hunk 7]
```diff
@@ -149,6 +140,7 @@ export class EmbeddedClientApplication {
             messageChannel.port1.onmessage = ((message: MessageEvent): void => {
                 this.logger.verbose(`in messageBroker<T> w/ origin: ${message}`);
                 clearTimeout(timeoutId);
+                messageChannel.port1.close();
```
Comment (sameerag) [inline lib/msal-browser/src/broker/EmbeddedClientApplication.ts:143]: port1 and port2, if they have significance, can we please add comments?

**Review by sameerag — APPROVED** [review]
Some comments.

File: lib/msal-browser/src/broker/EmbeddedClientApplication.ts (hunk 7 above)
Comment (tnorling) [reply lib/msal-browser/src/broker/EmbeddedClientApplication.ts:147]: This is part of the MessageChannel API and is documented on MDN. I did also add a brief comment to explain what this is.

File: lib/msal-browser/src/broker/BrokerAuthResponse.ts (hunk 6 above)
Comment (tnorling) [reply lib/msal-browser/src/broker/BrokerAuthResponse.ts]: Done
